lib/Conversion/GPUToROCM/ConvertKernelFuncToHSACO.cpp

```diff
@@ -0,0 +1,407 @@
+
+  std::string errorMessage;
+  int lldResult = llvm::sys::ExecuteAndWait(
+      lldProgram, llvm::ArrayRef<llvm::StringRef>(lldArgs), llvm::None, {}, 0,
+      0, &errorMessage);
+  if (lldResult) {
+    module.emitError("ld.lld execution failed : " + errorMessage);
+    return {};
+  }
+  // read HSACO
+  auto hsacoFileOrError = llvm::MemoryBuffer::getFileAsStream(hsacoPath);
+  if ((ec = hsacoFileOrError.getError()))
+    return {};
+
+  std::unique_ptr<llvm::MemoryBuffer> hsacoFile =
+      std::move(hsacoFileOrError.get());
+
+  return std::make_unique<std::vector<char>>(hsacoFile->getBufferStart(),
+                                             hsacoFile->getBufferEnd());
+}
+
+OwnedHSACO
+GpuKernelToHSACOPass::emitModuleToHSACOForTesting(llvm::Module &llvmModule,
+                                                  ModuleOp module) {
+  const char data[] = "HSACO";
+  return std::make_unique<std::vector<char>>(data, data + sizeof(data) - 1);
+}
+
+OwnedHSACO GpuKernelToHSACOPass::convertModuleToHSACO(llvm::Module &llvmModule,
+                                                      ModuleOp module) {
+  if (config.testMode)
+    return emitModuleToHSACOForTesting(llvmModule, module);
+
+  // Construct LLVM TargetMachine for AMDGPU target.
+  std::unique_ptr<llvm::TargetMachine> targetMachine;
+  {
+    std::string error;
+    llvm::Triple triple("amdgcn--amdhsa-amdgiz");
+    const llvm::Target *target =
+        llvm::TargetRegistry::lookupTarget("", triple, error);
+    if (target == nullptr) {
+      module.emitError("Cannot initialize target triple");
+      return {};
+    }
+    std::string mcpuStr = getMcpuOptionString(config.amdgpuVersion);
+    std::string codeObjectStr = getCodeObjectOptionString(config.hsacoVersion);
+    targetMachine.reset(target->createTargetMachine(triple.str(), mcpuStr,
+                                                    codeObjectStr, {}, {}));
+  }
+
+  // Set the data layout of the llvm module to match what the target needs.
+  llvmModule.setDataLayout(targetMachine->createDataLayout());
+
+  if (failed(linkROCDLIfNecessary(llvmModule, module, config.amdgpuVersion,
+                                  config.rocdlDir)))
+    return {};
+
+  // Lower LLVM module to HSA code object
+  return emitModuleToHSACO(llvmModule, module, *targetMachine);
+}
+
+StringAttr GpuKernelToHSACOPass::translateGpuModuleToHSACOAnnotation(
+    llvm::Module &llvmModule, ModuleOp module) {
+
+  OwnedHSACO hsaco = convertModuleToHSACO(llvmModule, module);
+  if (!hsaco)
+    return {};
+
+  return StringAttr::get({hsaco->data(), hsaco->size()}, module.getContext());
+}
+
+std::unique_ptr<OpPassBase<ModuleOp>> mlir::createConvertGPUKernelToHSACOPass(
+    rocm::HSACOGeneratorConfig hsacoGeneratorConfig) {
+  return std::make_unique<GpuKernelToHSACOPass>(hsacoGeneratorConfig);
+}
+
+static PassRegistration<GpuKernelToHSACOPass>
+    pass("test-kernel-to-hsaco",
+         "Convert all kernel functions to HSA code object blobs");
```

Comment:
With a macro that define if ROCM is installed or not, we can then conditionally register another version of the pass "rocm-hsaco-lowering" or something like which actually run the linker as well, and add a test for this.